File: code/Python/Chapter8/8-7/quick_sort.py
import random

INSERTION_SORT_THRESHOLD = 10

def insertion_sort(lst):
    """Binary Insertion Sort"""
    for i in range(1, len(lst)):
        temp = lst[i]
        left = 0
        right = i - 1

        while left <= right:
            mid = (left + right) // 2
            if temp < lst[mid]:
                right = mid - 1
            else:
                left = mid + 1

        for j in range(i - 1, left - 1, -1):
            lst[j + 1] = lst[j]
        lst[left] = temp
    
    return lst

def partition(lst, start, end, pivot):
    i = start + 1
    j = end
    while i <= j:
        while i <= j and lst[i] <= pivot:
            i += 1
        while i <= j and lst[j] >= pivot:
            j -= 1
        if i <= j:
            lst[i], lst[j] = lst[j], lst[i]
    
    lst[start], lst[j] = lst[j], lst[start]
    return j
    
def median_of_three(lst, start, end):
    mid = start + (end - start) // 2

    if lst[start] > lst[mid]:
        lst[start], lst[mid] = lst[mid], lst[start]
    if lst[start] > lst[end]:
        lst[start], lst[end] = lst[end], lst[start]
    if lst[mid] > lst[end]:
        lst[mid], lst[end] = lst[end], lst[mid]
    
    return mid

def quick_sort(lst):
    """Quick Sort (original v1.0)"""
    def sort(start, end):
        if start < end:
            pivot = lst[start]

            pivot_index = partition(lst, start, end, pivot)
            sort(start, pivot_index - 1)
            sort(pivot_index + 1, end)

    sort(0, len(lst) - 1)
    return lst

# def quick_sort(lst):
#     """Quick Sort (optimized v2.0)"""
#     def sort(start, end):
#         if start < end:
#             pivot_index = random.randint(start, end)
#             lst[start], lst[pivot_index] = lst[pivot_index], lst[start]
#             pivot = lst[start]

#             pivot_index = partition(lst, start, end, pivot)
#             sort(start, pivot_index - 1)
#             sort(pivot_index + 1, end)

#     sort(0, len(lst) - 1)
#     return lst

# def quick_sort(lst):
#     """Quick Sort (optimized v2.1)"""
#     def sort(start, end):
#         if start < end:
#             pivot_index = median_of_three(lst, start, end)
#             lst[start], lst[pivot_index] = lst[pivot_index], lst[start]
#             pivot = lst[start]

#             pivot_index = partition(lst, start, end, pivot)
#             sort(start, pivot_index - 1)
#             sort(pivot_index + 1, end)

#     sort(0, len(lst) - 1)
#     return lst

# def quick_sort(lst):
#     """Quick Sort (optimized v2.2)"""
#     def sort(start, end):
#         if end - start <= INSERTION_SORT_THRESHOLD:
#             lst[start:end + 1] = insertion_sort(lst[start:end + 1])
#             return

#         pivot_index = median_of_three(lst, start, end)
#         lst[start], lst[pivot_index] = lst[pivot_index], lst[start]
#         pivot = lst[start]

#         pivot_index = partition(lst, start, end, pivot)
#         sort(start, pivot_index - 1)
#         sort(pivot_index + 1, end)

#     sort(0, len(lst) - 1)
#     return lst

lst = [4, 7, 6, 5, 3, 2, 8, 1]
quick_sort(lst)
print(lst)

File: code/Java/Chapter8/8-7/QuickSort.java
import java.util.Arrays;
import java.util.Random;

public class QuickSort {
    private static final int INSERTION_SORT_THRESHOLD = 10;

    public static void main(String[] args) {
        int[] arr = {4, 7, 6, 5, 3, 2, 8, 1};
        quickSort(arr);
        System.out.println(Arrays.toString(arr));
    }

    /**
     * Binary Insertion Sort
     */
    private static int[] insertionSort(int[] arr) {
        for (int i = 1; i < arr.length; i++) {
            int temp = arr[i];
            int left = 0;
            int right = i - 1;

            while (left <= right) {
                int mid = (left + right) / 2;
                if (temp < arr[mid]) {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            }

            for (int j = i - 1; j >= left; j--) {
                arr[j + 1] = arr[j];
            }
            arr[left] = temp;
        }

        return arr;
    }

    private static int partition(int[] arr, int start, int end, int pivot) {
        int i = start + 1;
        int j = end;

        while (i <= j) {
            while (i <= j && arr[i] <= pivot) {
                i++;
            }
            while (i <= j && arr[j] >= pivot) {
                j--;
            }
            if (i <= j) {
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        int temp = arr[start];
        arr[start] = arr[j];
        arr[j] = temp;
        return j;
    }

    private static int medianOfThree(int[] arr, int start, int end) {
        int mid = start + (end - start) / 2;

        int temp;
        if (arr[start] > arr[mid]) {
            temp = arr[start];
            arr[start] = arr[mid];
            arr[mid] = temp;
        }
        if (arr[start] > arr[end]) {
            temp = arr[start];
            arr[start] = arr[end];
            arr[end] = temp;
        }
        if (arr[mid] > arr[end]) {
            temp = arr[mid];
            arr[mid] = arr[end];
            arr[end] = temp;
        }

        return mid;
    }

    /**
     * Quick Sort
     */
    public static int[] quickSort(int[] arr) {
        quickSort(arr, 0, arr.length - 1);
        return arr;
    }

    /**
     * Help function for Quick Sort (original v1.0)
     */
    private static void quickSort(int[] arr, int start, int end) {
        if (start < end) {
            int pivot = arr[start];

            int pivot_index = partition(arr, start, end, pivot);
            quickSort(arr, start, pivot_index - 1);
            quickSort(arr, pivot_index + 1, end);
        }
    }

    // /**
    //  * Help function for Quick Sort (optimized v2.0)
    //  */
    // private static void quickSort(int[] arr, int start, int end) {
    //     if (start < end) {
    //         Random random = new Random();
    //         int pivot_index = random.nextInt(end - start + 1) + start;
    //         int temp = arr[start];
    //         arr[start] = arr[pivot_index];
    //         arr[pivot_index] = temp;
    //         int pivot = arr[start];

    //         pivot_index = partition(arr, start, end, pivot);
    //         quickSort(arr, start, pivot_index - 1);
    //         quickSort(arr, pivot_index + 1, end);
    //     }
    // }

    // /**
    //  * Help function for Quick Sort (optimized v2.1)
    //  */
    // private static void quickSort(int[] arr, int start, int end) {
    //     if (start < end) {
    //         int pivot_index = medianOfThree(arr, start, end);
    //         int temp = arr[start];
    //         arr[start] = arr[pivot_index];
    //         arr[pivot_index] = temp;
    //         int pivot = arr[start];

    //         pivot_index = partition(arr, start, end, pivot);
    //         quickSort(arr, start, pivot_index - 1);
    //         quickSort(arr, pivot_index + 1, end);
    //     }
    // }

    // /**
    //  * Help function for Quick Sort (optimized v2.2)
    //  */
    // private static void quickSort(int[] arr, int start, int end) {
    //     if (end - start <= INSERTION_SORT_THRESHOLD) {
    //         int[] temp = new int[end - start + 1];
    //         System.arraycopy(arr, start, temp, 0, end - start + 1);
    //         insertionSort(temp);
    //         System.arraycopy(temp, 0, arr, start, end - start + 1);
    //         return;
    //     }

    //     int pivot_index = medianOfThree(arr, start, end);
    //     int temp = arr[start];
    //     arr[start] = arr[pivot_index];
    //     arr[pivot_index] = temp;
    //     int pivot = arr[start];

    //     pivot_index = partition(arr, start, end, pivot);
    //     quickSort(arr, start, pivot_index - 1);
    //     quickSort(arr, pivot_index + 1, end);
    // }
}
